pynisher/pynisher.py

```diff
@@ -0,0 +1,450 @@
+        limiter = Limiter.create(
+            func=self.func,
+            output=send_pipe,
+            memory=self.memory,
+            cpu_time=self.cpu_time,
+            warnings=self.warnings,
+        )
+
+        # We now create the subprocess and let it know that it should call the limiter's
+        # __call__ with the args and kwargs for the function being limited
+        subprocess = self.context.Process(
+            target=limiter.__call__,
+            args=args,
+            kwargs=kwargs,
+            daemon=False,
+            name=self.name,
+        )
+
+        # 4 kinds of `response` we expect
+        #
+        # * (result, None, None)    | success
+        #                               => No error, we got a result
+        # * (None, error, traceback)| failed
+        #                               =>  Error, CPUtimeout linux/mac, MemoryError
+        # * None                    | failed
+        #                               => MemoryError during sending of real error
+        # * EMPTY                   | failed, nothing received from pipe
+        #                               => Walltime, CPUTimeout windows, Unknown
+        result = EMPTY
+        err: Exception | None = None
+        tb: str | None = None
+
+        # Let loose
+        subprocess.start()
+
+        # If self.wall time is None, block until the subprocess finishes or terminates
+        # Otherwise, will return after wall_time and the process will still be running
+        subprocess.join(self.wall_time)
+
+        # exitcode here can only take on 3 values
+        #
+        # * None        | the subprocess is still running (walltime elapsed)
+        # * 0           | Ended gracefully, pipe is closed or response in the pipe
+        # * >0          | Process was terminated non-gracefully, nothing in the pipe
+        #               | and it may not be closed
+        exitcode = subprocess.exitcode
+
+        # Wall time expired
+        if exitcode is None:
+            subprocess.terminate()
+
+            result = EMPTY
+            err = WallTimeoutException(
+                f"Did not finish in time ({self.wall_time}s)"
+                f"\n{callstring(self.func, *args, **kwargs)}"
+            )
+
+        # Completed gracefully
+        elif exitcode == 0:
+
+            # Pipe has data in it or has been closed properly
+            if receive_pipe.poll():
+
+                # If there is data, it will be read
+                try:
+                    response = receive_pipe.recv()
+                    if response is None:
+                        result = EMPTY
+                        err = MemoryLimitException(
+                            "Ended gracefully but only `None` could be sent back"
+                            " from the process. This is a MemoryException as"
+                            " previous errors could not be sent previously."
```

Comment:
I wouldn't understand this error message as a user.

Reply:
Makes sense, Ill fix